libcudacxx/include/cuda/__iterator/transform_iterator.h

```diff
@@ -0,0 +1,501 @@
+  {
+    return __lhs.__current_ <=> __rhs.__current_;
+  }
+#endif // !_LIBCUDACXX_HAS_NO_SPACESHIP_OPERATOR
+
+  //! @brief Returns a copy of the \c transform_iterator \p __i advanced by \p __n
+  //! @param __i The \c transform_iterator to advance
+  //! @param __n The number of elements to advance
+  template <class _Iter2 = _Iter>
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr auto
+  operator+(const transform_iterator& __i, difference_type __n)
+    _CCCL_TRAILING_REQUIRES(transform_iterator)(_CUDA_VSTD::random_access_iterator<_Iter2>)
+  {
+    return transform_iterator{__i.__current_ + __n, *__i.__func_};
+  }
+
+  //! @brief Returns a copy of the \c transform_iterator \p __i advanced by \p __n
+  //! @param __n The number of elements to advance
+  //! @param __i The \c transform_iterator to advance
+  template <class _Iter2 = _Iter>
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr auto
+  operator+(difference_type __n, const transform_iterator& __i)
+    _CCCL_TRAILING_REQUIRES(transform_iterator)(_CUDA_VSTD::random_access_iterator<_Iter2>)
+  {
+    return transform_iterator{__i.__current_ + __n, *__i.__func_};
+  }
+
+  //! @brief Returns a copy of the the \c transform_iterator \p __i decremented by \p __n
+  //! @param __i The \c transform_iterator to decrement
+  //! @param __n The number of elements to decrement
+  template <class _Iter2 = _Iter>
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr auto
+  operator-(const transform_iterator& __i, difference_type __n)
+    _CCCL_TRAILING_REQUIRES(transform_iterator)(_CUDA_VSTD::random_access_iterator<_Iter2>)
+  {
+    return transform_iterator{__i.__current_ - __n, *__i.__func_};
+  }
+
+  //! @brief Returns the distance between \p __lhs and \p __rhs
+  //! @param __lhs The left \c transform_iterator
+  //! @param __rhs The right \c transform_iterator
+  //! @return The distance between the stored iterators
+  template <class _Iter2 = _Iter>
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI friend constexpr auto
+  operator-(const transform_iterator& __lhs, const transform_iterator& __rhs)
+    _CCCL_TRAILING_REQUIRES(difference_type)(_CUDA_VSTD::sized_sentinel_for<_Iter2, _Iter2>)
+  {
+    return __lhs.__current_ - __rhs.__current_;
+  }
+};
+
+//! @brief make_transform_iterator creates a \p transform_iterator from an \c _Iter and a \c _Fn.
+//!
+//! @param __iter The \c Iterator pointing to the input range of the newly created \p transform_iterator.
+//! @param __fun The \c _Fn used to transform the range pointed to by @param __iter in the newly created
+//! \p transform_iterator.
+//! @return A new \p transform_iterator which transforms the range at @param __iter by @param __fun.
+template <class _Iter, class _Fn>
+[[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr auto make_transform_iterator(_Iter __iter, _Fn __fun)
+{
+  return transform_iterator<_Iter, _Fn>{__iter, __fun};
+}
```

Comment:
Suggestion: I strongly believe we should rely on CTAD and no longer provide `make_xxx` functions. I would remove it.

Reply:
I am all for it, but the goal is to have a drop in replacement for `thrust::transfrom_iterator`

Reply:
The Thrust version btw, has a different order of template parameters:
https://github.com/NVIDIA/cccl/blob/cd6a09097f8b92a5f81e625c0ffd2f56fda24d00/thrust/thrust/iterator/transform_iterator.h#L353-L355

Reply:
That one has annoyed me, because the order of the function arguments is different than the order of template arguments.

I opted to change the order of the template arguments

Reply:
I would have also intuitively picked your order.